src/test/java/com/neo4j/docker/coredb/TestBasic.java

```diff
@@ -187,11 +197,36 @@ void testShutsDownCleanly(String signal) throws Exception
     @Test
     void testStartsWhenDebuggingEnabled()
     {
-        try(GenericContainer container = createBasicContainer())
+        try ( GenericContainer container = createBasicContainer() )
         {
             container.withEnv( "NEO4J_DEBUG", "true" );
             container.start();
             Assertions.assertTrue( container.isRunning() );
         }
     }
+
+    @Test
+    void testContainerCanBeRestarted() throws IOException
+    {
+        try ( GenericContainer container = createBasicContainer() )
+        {
+            int boltHostPort = getUniqueHostPort();
+            int browserHostPort = getUniqueHostPort();
+
+            container.waitingFor( waitForBoltReady( Duration.ofSeconds( 90 ) ) );
+            container.withEnv( "NEO4J_AUTH", "none" );
+
+            container.setPortBindings( List.of( browserHostPort + ":7474", boltHostPort + ":7687" ) );
+
+            container.start();
+
+            log.info( "Terminating container with SIGKILL signal" );
+            container.getDockerClient().killContainerCmd( container.getContainerId() ).withSignal( "SIGKILL" ).exec();
+
+            log.info( "Starting container" );
+            container.getDockerClient().startContainerCmd( container.getContainerId() ).exec();
+
+            waitForBoltReady( Duration.ofSeconds( 90 ) ).waitUntilReady( container );
```

Comment:
This is some really great refactoring, I like what you've done a lot :+1: 

My only suggestion would be that, in 6 months time I'm going to look at this and think "Why is it allocating ports this complicated way?" "Why is it using `DockerClient` to start/stop the container?" (I know you are sufficiently traumatised by this bug that you probably won't have forgotten by then :grin: ). In situations where I have to write code that does stuff in a weird non-obvious way, I find it's good practice to write a comment about **why** it's doing the weird thing. Mainly to save my future self from confusion.

Anyway all that is to say, could you add a comment somewhere about how this is a unit test to replecate `docker-desktop` going away while a neo4j container is running, and another comment about how setting the ports manually means that TestContainers doesn't reassign ports and break the test.